app/Libraries/Database.php

```diff
@@ -39,41 +40,56 @@ class Database
   public static $error;
 
   /**
-   * Database constructor
-   *
+   * @var Cookie
    */
-  private function __construct($credentials)
-  {
-    if (Cookie::exists('user_id')) {
-      $user = UserStorage::getUserById(Cookie::get('user_id'));
-      $credentials = (array) $user;
-    }
+  public $cookie;
 
-    $dsn = "mysql:host=" . $credentials['host'];
-    $options = array(
-      PDO::ATTR_PERSISTENT => true,
-      PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION
-    );
+  /**
+   * @var UserStorage
+   */
+  public $storage;
 
-    try {
-      self::$pdo = new PDO($dsn, $credentials['username'], $credentials['password'], $options);
-    } catch (PDOException $e) {
-      throw $e;
-    }
-  }
+  /**
+   * @var Request
+   */
+  public $request;
 
   /**
-   * Get connection instance
+   * Database Constructor
    *
-   * @return array
+   * @param Cookie $cookie
+   * @param UserStorage $storage
+   * @param Request $request
    */
-  public static function getInstance(array $credentials = []): object
+  public function __construct(Cookie $cookie, UserStorage $storage, Request $request)
   {
-    if (self::$instance == null) {
-      self::$instance = new Database($credentials);
-    }
+    $this->cookie = $cookie;
+    $this->storage = $storage;
+    $this->request = $request;
+    $this->connect();
+  }
 
-    return self::$instance;
+  public function connect($credentials = null)
+  {
+    if(!$credentials && !$this->cookie->exists('user_id')) return;
+
+    if ($this->cookie->exists('user_id')) {
+      $user = $this->storage->getUserById($this->cookie->get('user_id'));
+      $credentials = (array) $user;
+    }
+  
+    try {
+      $dsn = "mysql:host=" . $credentials['host'];
+      $options = array(
+        PDO::ATTR_PERSISTENT => true,
+        PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION
+      );
+
+      $this->pdo = new PDO($dsn, $credentials['username'], $credentials['password'], $options);
+      return $this->pdo;
+    } catch (PDOException $e) {

```

Comment:
Why are you catching the exception just to rethrow it ? Maybe its better to just remove the try catch. Or if you want to catch it, log it using the logger and then throw other custom Database exception

Reply:
I have a try/catch block because I need to handle the scenario when an user enters invalid credentials. Hence I don't want to have extra logic in here that's why I rethrow the exception and therefore I'm keeping the connect method clean. ( Logging and Session messages are handled in Login.php Controller ).